core/src/main/java/com/datastax/oss/driver/internal/core/ssl/SniSslEngineFactory.java

```diff
@@ -44,17 +46,40 @@ public SSLEngine newSslEngine(@NonNull EndPoint remoteEndpoint) {
               "Configuration error: can only use %s with SNI end points",
               this.getClass().getSimpleName()));
     }
-    SSLEngine engine;
     SniEndPoint sniEndPoint = (SniEndPoint) remoteEndpoint;
+    InetSocketAddress address = sniEndPoint.resolve();
+    String sniServerName = sniEndPoint.getServerName();
 
-    engine = sslContext.createSSLEngine();
+    // When hostname verification is enabled (with setEndpointIdentificationAlgorithm), the SSL
+    // engine will try to match the server's certificate against the SNI host name; if that doesn't
+    // work, it will fall back to the "advisory peer host" passed to createSSLEngine.
+    //
+    // In our case, the first check will never succeed because our SNI host name is not the DNS name
+    // (we use the Cassandra host_id instead). So we *must* set the advisory peer information.
+    //
+    // However if we use the address as-is, this leads to another issue: the advisory peer
+    // information is also used to cache SSL sessions internally. All of our nodes share the same
+    // proxy address, so the JDK tries to reuse SSL sessions across nodes. But it doesn't update the
+    // SNI host name every time, so it ends up opening connections to the wrong node.
+    //
+    // To avoid that, we create a unique "fake" port for every node. We still get session reuse for
+    // a given node, but not across nodes. This is safe because the advisory port is only used for
+    // session caching.
+    SSLEngine engine =
+        sslContext.createSSLEngine(address.getHostName(), getFakePort(sniServerName));
     engine.setUseClientMode(true);
     SSLParameters parameters = engine.getSSLParameters();
-    parameters.setServerNames(ImmutableList.of(new SNIHostName(sniEndPoint.getServerName())));
+    parameters.setServerNames(ImmutableList.of(new SNIHostName(sniServerName)));
+    parameters.setEndpointIdentificationAlgorithm("HTTPS");
     engine.setSSLParameters(parameters);
     return engine;
   }
 
+  private int getFakePort(String sniServerName) {
+    fakePorts.addIfAbsent(sniServerName);
+    return fakePorts.indexOf(sniServerName);
```

Comment:
The fake port's actual number doesn't seem to make any difference, so instead of storing host names in a list, why don't you use an `AtomicInteger`, preferably starting at the beginning of the ephemeral port range (49152)? What makes me a bit uncomfortable is that zero has a special meaning when used as a port, and low numbers also usually mean reserved ports. You stated that the number is really only used for session caching, but who knows if this won't change in future JDK versions.

Reply:
OK I saw your comments regarding using AtomicInteger. Please disregard this comment. I only hope using zero as a port number won't bite us back one day.

Reply:
@adutra could you tell me what was the reason?

Reply:
> I only hope using zero as a port number won't bite us back one day.

We can always add a constant offset to the list index.

Reply:
> starting at the beginning of the ephemeral port range (49152)

I missed that the first time I read your comment, sounds perfect.